sound/soc/sof/intel/hda-loader.c

```diff
@@ -556,16 +587,28 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
 	stripped_firmware.data = fw_lib->sof_fw.fw->data + fw_lib->sof_fw.payload_offset;
 	stripped_firmware.size = fw_lib->sof_fw.fw->size - fw_lib->sof_fw.payload_offset;
 
+	/*
+	 * force re-allocation of the cl_dmab if the preserved DMA buffer is
+	 * smaller than what is needed for the library
+	 */
+	if (persistent_cl_buffer && stripped_firmware.size > hda->cl_dmab.bytes) {
+		snd_dma_free_pages(&hda->cl_dmab);
+		hda->cl_dmab.area = NULL;
+		hda->cl_dmab.bytes = 0;
```

Comment:
When will cl_dmab be allocated?

Reply:
hda_cl_prepare() allocates the buffer if it is not yet allocated.
If the buffer which we allocated for the basefw is smaller than what we would need for a library, we free the allocated buffer and let hda_cl_prepare() to re-allocate a new, bigger buffer - which will be used for fw download as well later.